File: common/src/main/java/cn/breadnicecat/candycraftce/utils/CLogUtils.java
package cn.breadnicecat.candycraftce.utils;

import cn.breadnicecat.candycraftce.CandyCraftCE;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;

import static cn.breadnicecat.candycraftce.utils.CommonUtils.getCaller;

/**
 * Created in 2023/8/22 21:58
 * Project: candycraftce
 *
 * @author <a href="https://github.com/Bread-Nicecat">Bread_NiceCat</a>
 */
public class CLogUtils {
	private static final HashMap<String, Logger> CACHE = new HashMap<>();
	private static final HashMap<Class<?>, Logger> CACHE_BY_CLASS = new HashMap<>();
	private static final HashMap<Class<?>, Logger> SIGN_VALIDATOR = new HashMap<>();
	
	public static Logger getModLogger() {
		return getModLogger(getCaller());
	}
	
	public static Logger getModLogger(Class<?> clazz) {
		return CACHE_BY_CLASS.computeIfAbsent(clazz, (k) -> getModLogger(k.getSimpleName()));
	}
	
	
	public static Logger getModLogger(String name) {
		String id = CandyCraftCE.MOD_NAME + "/" + name;
		return CACHE.computeIfAbsent(id, LoggerFactory::getLogger);
	}
	
	/**
	 * Should invoke in all {@code init()}
	 */
	@NotNull
	public static Logger sign() {
		Class<?> caller = getCaller();
		boolean dup = SIGN_VALIDATOR.containsKey(caller);
		Logger l = getModLogger(caller);
		l.info("{} signed", caller.getName());
		if (dup) {
			l.warn("Duplicate sign for {}", caller.getName());
		} else {
			SIGN_VALIDATOR.put(caller, l);
		}
		return l;
	}
	
}

File: src/main/kotlin/cn/breadnicecat/candycraftce/utils/CLogUtils.kt
package cn.breadnicecat.candycraftce.utils

import net.minecraft.resources.ResourceLocation
import org.slf4j.Logger
import org.slf4j.LoggerFactory

/**
 * Created by NiceCat on 2026/5/1.
 * Project: candycraftce
 * @author <a href="https://github.com/BreadNiceCat">Bread_NiceCat</a>
 *
 */
object CLogUtils {

    val walker: StackWalker = StackWalker.getInstance(StackWalker.Option.RETAIN_CLASS_REFERENCE)

    private val logCache = mutableMapOf<String, Logger>()
    val mainLog = modLogger("Core")
    val registerLog = modLogger("Registry")
    val debugLog = modLogger("Debug")
    val clog: Logger
        get() = modLogger(walker.callerClass.let {
            if (it.kotlin.isCompanion) it.enclosingClass else it
        }.simpleName)

    fun modLogger(tag: String): Logger {
        return logCache.computeIfAbsent(tag) { LoggerFactory.getLogger("CandyCraftCE|${it}") }
    }

    private val signed = mutableSetOf<Class<*>>()
    private var late = false
    fun sign() {
        val clazz = walker.callerClass
        require(signed.add(clazz)) { "Class ${clazz.simpleName} is already registered" }
        mainLog.info("${clazz.simpleName} loaded")
        if (late) {
            mainLog.error("This class is late for sign! It is a bug, please report it!")
        }
    }

    internal fun markLateForSign() {
        late = true
    }

    fun logRegister(type: String, id: ResourceLocation) {
        registerLog.info("Registering $type/$id")
    }
}
